src/async-wrap.cc

```diff
@@ -617,10 +637,29 @@ AsyncWrap::AsyncWrap(Environment* env,
 
   // Use AsyncReset() call to execute the init() callbacks.
   AsyncReset(silent);
+
+  switch (provider_type()) {
+#define V(PROVIDER)                                                           \
+    case PROVIDER_ ## PROVIDER:                                               \
+      TRACE_EVENT_NESTABLE_ASYNC_BEGIN0("node", #PROVIDER,                    \
+        static_cast<int64_t>(get_id()));                                      \
+      break;
+    NODE_ASYNC_PROVIDER_TYPES(V)
+#undef V
+  }
 }
 
 
 AsyncWrap::~AsyncWrap() {
+  switch (provider_type()) {
+#define V(PROVIDER)                                                           \
```

Comment:
Here are places where this will fail:

1. `HTTPParser` when `require('_http_common').freeParser()` is called. An internal pool of parsers is created and every time one is used it's assigned a new async id. To get around this `TRACE_EVENT_NESTABLE_ASYNC_BEGIN0` needs to be placed in `Parser::Reinitialize()` (`src/node_http_parser.cc`) and `TRACE_EVENT_NESTABLE_ASYNC_END0` at the `emitDestroy()` call in `freeParser()` (`lib/_http_common.js`).
2. Every `Timeout` and `Immediate`, since they're all pseudo async resources (depending on the `TimerWrap` to fire). Since they depend on the `TimerWrap` there will still be an associated async id, just not the greater detail of each JS object.

Neither of these would be a blocker for me landing this PR. Just note that (1) will not have the same async ids going in as it does going out. Also I would suggest we exclude `nextTick()` calls.